File: ios/EGETrainerApp/EGETrainerApp/Models/TeacherModels.swift
import Foundation

// MARK: - list_my_students

struct StudentListItem: Decodable, Identifiable, Hashable {
    var studentId: String
    var email: String?
    var firstName: String?
    var lastName: String?
    var studentGrade: Int?
    var linkedAt: String?

    enum CodingKeys: String, CodingKey {
        case email
        case studentId = "student_id"
        case firstName = "first_name"
        case lastName = "last_name"
        case studentGrade = "student_grade"
        case linkedAt = "linked_at"
    }

    var id: String { studentId }

    var displayName: String {
        let name = [firstName, lastName].compactMap { $0 }.joined(separator: " ")
        if !name.trimmingCharacters(in: .whitespaces).isEmpty { return name }
        return email ?? "Ученик"
    }
}

// MARK: - teacher_students_summary

struct StudentSummary: Decodable {
    var studentId: String
    var lastSeenAt: String?
    var activityTotal: Int?
    var last10Total: Int?
    var last10Correct: Int?
    var coveredTopicsAllTime: Int?

    enum CodingKeys: String, CodingKey {
        case studentId = "student_id"
        case lastSeenAt = "last_seen_at"
        case activityTotal = "activity_total"
        case last10Total = "last10_total"
        case last10Correct = "last10_correct"
        case coveredTopicsAllTime = "covered_topics_all_time"
    }
}

// MARK: - consent: запросы учителя (исходящие) и ученика (входящие)

struct OutgoingStudentRequest: Decodable, Identifiable {
    var requestId: String
    var studentEmail: String?
    var status: String?
    var createdAt: String?

    enum CodingKeys: String, CodingKey {
        case status
        case requestId = "request_id"
        case studentEmail = "student_email"
        case createdAt = "created_at"
    }

    var id: String { requestId }
}

struct IncomingTeacherRequest: Decodable, Identifiable {
    var requestId: String
    var teacherId: String?
    var teacherName: String?
    var teacherEmail: String?
    var createdAt: String?
    var status: String?

    enum CodingKeys: String, CodingKey {
        case status
        case requestId = "request_id"
        case teacherId = "teacher_id"
        case teacherName = "teacher_name"
        case teacherEmail = "teacher_email"
        case createdAt = "created_at"
    }

    var id: String { requestId }
}

struct MyTeacher: Decodable, Identifiable {
    var teacherId: String
    var teacherName: String?
    var teacherEmail: String?
    var linkedAt: String?

    enum CodingKeys: String, CodingKey {
        case teacherId = "teacher_id"
        case teacherName = "teacher_name"
        case teacherEmail = "teacher_email"
        case linkedAt = "linked_at"
    }

    var id: String { teacherId }
    var displayName: String { teacherName ?? teacherEmail ?? "Преподаватель" }
}

// MARK: - list_student_attempts

struct StudentAttemptRow: Decodable, Identifiable {
    var attemptId: String
    var homeworkId: String?
    var homeworkTitle: String?
    var total: Int?
    var correct: Int?
    var startedAt: String?
    var finishedAt: String?
    var durationMs: Int?

    enum CodingKeys: String, CodingKey {
        case total, correct
        case attemptId = "attempt_id"
        case homeworkId = "homework_id"
        case homeworkTitle = "homework_title"
        case startedAt = "started_at"
        case finishedAt = "finished_at"
        case durationMs = "duration_ms"
    }

    var id: String { attemptId }
}

// MARK: - teacher_picking_screen_v2

struct PickingScreen: Decodable {
    var screen: PickingScreenMeta?
    var sections: [PickSection]?
    var student: PickingStudent?

    struct PickingScreenMeta: Decodable {
        var mode: String?
        var canPick: Bool?
        var sessionSeed: String?
        var supportedFilters: [String]?

        enum CodingKeys: String, CodingKey {
            case mode
            case canPick = "can_pick"
            case sessionSeed = "session_seed"
            case supportedFilters = "supported_filters"
        }
    }

    struct PickingStudent: Decodable {
        var days: Int?
        var source: String?
        var studentId: String?

        enum CodingKeys: String, CodingKey {
            case days, source
            case studentId = "student_id"
        }
    }
}

struct PickSection: Decodable, Identifiable {
    var sectionId: String
    var title: String?
    var sortOrder: Int?
    var filterCounts: FilterCounts?
    var topics: [PickTopic]?

    enum CodingKeys: String, CodingKey {
        case title, topics
        case sectionId = "section_id"
        case sortOrder = "sort_order"
        case filterCounts = "filter_counts"
    }

    var id: String { sectionId }
}

struct PickTopic: Decodable, Identifiable {
    var topicId: String
    var title: String?
    var sortOrder: Int?
    var coverage: PickCoverage?
    var progress: PickProgress?
    var topicState: PickTopicState?
    var filterCounts: FilterCounts?

    enum CodingKeys: String, CodingKey {
        case title, coverage, progress
        case topicId = "topic_id"
        case sortOrder = "sort_order"
        case topicState = "topic_state"
        case filterCounts = "filter_counts"
    }

    var id: String { topicId }
}

struct PickCoverage: Decodable {
    var totalProtoCount: Int?
    var coveredProtoCount: Int?

    enum CodingKeys: String, CodingKey {
        case totalProtoCount = "total_proto_count"
        case coveredProtoCount = "covered_proto_count"
    }
}

struct PickProgress: Decodable {
    var allTimePct: Int?
    var lastSeenAt: String?
    var attemptCountTotal: Int?
    var correctCountTotal: Int?
    var subtopicLast3AvgPct: Double?

    enum CodingKeys: String, CodingKey {
        case allTimePct = "all_time_pct"
        case lastSeenAt = "last_seen_at"
        case attemptCountTotal = "attempt_count_total"
        case correctCountTotal = "correct_count_total"
        case subtopicLast3AvgPct = "subtopic_last3_avg_pct"
    }
}

struct PickTopicState: Decodable {
    var isStale: Bool?
    var isLowSeen: Bool?
    var isNotSeen: Bool?
    var isUnstable: Bool?

    enum CodingKeys: String, CodingKey {
        case isStale = "is_stale"
        case isLowSeen = "is_low_seen"
        case isNotSeen = "is_not_seen"
        case isUnstable = "is_unstable"
    }
}

struct FilterCounts: Decodable {
    var stale: Int?
    var unstable: Int?
    var unseenLow: Int?
    var weakSpots: Int?

    enum CodingKeys: String, CodingKey {
        case stale, unstable
        case unseenLow = "unseen_low"
        case weakSpots = "weak_spots"
    }
}

// MARK: - teacher_picking_resolve_batch_v1

// MARK: - Статистика прототипов (proto_last3_*, question_stats_for_teacher_*)

/// Бейдж «последние 3 попытки» по базовому прототипу (unic_id);
/// self-версия дополнительно отдаёт total/correct/last_attempt_at.
struct ProtoLast3Stat: Decodable {
    var unicId: String
    var last3Total: Int? = nil
    var last3Correct: Int? = nil
    var total: Int? = nil
    var correct: Int? = nil
    var lastAttemptAt: String? = nil

    enum CodingKeys: String, CodingKey {
        case unicId = "unic_id"
        case last3Total = "last3_total"
        case last3Correct = "last3_correct"
        case total, correct
        case lastAttemptAt = "last_attempt_at"
    }
}

struct QuestionStat: Decodable {
    var questionId: String
    var total: Int?
    var correct: Int?
    var last3Total: Int?
    var last3Correct: Int?
    var lastAttemptAt: String?

    enum CodingKeys: String, CodingKey {
        case questionId = "question_id"
        case total, correct
        case last3Total = "last3_total"
        case last3Correct = "last3_correct"
        case lastAttemptAt = "last_attempt_at"
    }
}

struct ResolveBatchResult: Decodable {
    var pickedQuestions: [ResolvedQuestion]?
    var shortages: [ResolveShortage]?

    enum CodingKeys: String, CodingKey {
        case pickedQuestions = "picked_questions"
        case shortages
    }
}

struct ResolveShortage: Decodable {
    var scopeId: String?
    var isShortage: Bool?
    var requestedN: Int?
    var returnedN: Int?
    var message: String?

    enum CodingKeys: String, CodingKey {
        case message
        case scopeId = "scope_id"
        case isShortage = "is_shortage"
        case requestedN = "requested_n"
        case returnedN = "returned_n"
    }
}

/// Вопрос из resolve-батча. question_id + topic_id достаточно для
/// создания ДЗ (frozen_questions) и резолва текста через ContentService.
struct ResolvedQuestion: Decodable, Identifiable, Hashable {
    var questionId: String
    var topicId: String?
    var protoId: String?
    var manifestPath: String?
    /// Атрибуция бакета из ответа батча (для клиентской ротации по бакетам).
    var scopeKind: String?
    var scopeId: String?

    enum CodingKeys: String, CodingKey {
        case questionId = "question_id"
        case topicId = "topic_id"
        case protoId = "proto_id"
        case manifestPath = "manifest_path"
        case scopeKind = "scope_kind"
        case scopeId = "scope_id"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        questionId = try c.decode(String.self, forKey: .questionId)
        topicId = try c.decodeIfPresent(String.self, forKey: .topicId)
        protoId = try c.decodeIfPresent(String.self, forKey: .protoId)
        manifestPath = try c.decodeIfPresent(String.self, forKey: .manifestPath)
        scopeKind = try c.decodeIfPresent(String.self, forKey: .scopeKind)
        scopeId = try c.decodeIfPresent(String.self, forKey: .scopeId)
    }

    init(questionId: String, topicId: String?, protoId: String? = nil, manifestPath: String? = nil,
         scopeKind: String? = nil, scopeId: String? = nil) {
        self.questionId = questionId
        self.topicId = topicId
        self.protoId = protoId
        self.manifestPath = manifestPath
        self.scopeKind = scopeKind
        self.scopeId = scopeId
    }

    var id: String { questionId }
}

// MARK: - создание ДЗ (REST insert)

struct HomeworkRow: Decodable {
    var id: String
    var title: String?
}

struct HomeworkLinkRow: Decodable {
    var id: String?
    var token: String?
}

File: android/EGETrainerApp/core/src/main/kotlin/ru/egetrainer/core/models/TeacherModels.kt
package ru.egetrainer.core.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// MARK: list_my_students

@Serializable
data class StudentListItem(
    @SerialName("student_id") val studentId: String,
    val email: String? = null,
    @SerialName("first_name") val firstName: String? = null,
    @SerialName("last_name") val lastName: String? = null,
    @SerialName("student_grade") val studentGrade: Int? = null,
    @SerialName("linked_at") val linkedAt: String? = null,
) {
    val id: String get() = studentId

    val displayName: String
        get() {
            val name = listOfNotNull(firstName, lastName).joinToString(" ").trim()
            if (name.isNotEmpty()) return name
            return email ?: "Ученик"
        }
}

// MARK: teacher_students_summary

@Serializable
data class StudentSummary(
    @SerialName("student_id") val studentId: String,
    @SerialName("last_seen_at") val lastSeenAt: String? = null,
    @SerialName("activity_total") val activityTotal: Int? = null,
    @SerialName("last10_total") val last10Total: Int? = null,
    @SerialName("last10_correct") val last10Correct: Int? = null,
    @SerialName("covered_topics_all_time") val coveredTopicsAllTime: Int? = null,
)

// MARK: consent: запросы учителя (исходящие) и ученика (входящие)

@Serializable
data class OutgoingStudentRequest(
    @SerialName("request_id") val requestId: String,
    @SerialName("student_email") val studentEmail: String? = null,
    val status: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
) {
    val id: String get() = requestId
}

@Serializable
data class IncomingTeacherRequest(
    @SerialName("request_id") val requestId: String,
    @SerialName("teacher_id") val teacherId: String? = null,
    @SerialName("teacher_name") val teacherName: String? = null,
    @SerialName("teacher_email") val teacherEmail: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
    val status: String? = null,
) {
    val id: String get() = requestId
}

@Serializable
data class MyTeacher(
    @SerialName("teacher_id") val teacherId: String,
    @SerialName("teacher_name") val teacherName: String? = null,
    @SerialName("teacher_email") val teacherEmail: String? = null,
    @SerialName("linked_at") val linkedAt: String? = null,
) {
    val id: String get() = teacherId
    val displayName: String get() = teacherName ?: teacherEmail ?: "Преподаватель"
}

// MARK: list_student_attempts

@Serializable
data class StudentAttemptRow(
    @SerialName("attempt_id") val attemptId: String,
    @SerialName("homework_id") val homeworkId: String? = null,
    @SerialName("homework_title") val homeworkTitle: String? = null,
    val total: Int? = null,
    val correct: Int? = null,
    @SerialName("started_at") val startedAt: String? = null,
    @SerialName("finished_at") val finishedAt: String? = null,
    @SerialName("duration_ms") val durationMs: Int? = null,
) {
    val id: String get() = attemptId
}

// MARK: teacher_picking_screen_v2

@Serializable
data class PickingScreen(
    val screen: PickingScreenMeta? = null,
    val sections: List<PickSection>? = null,
    val student: PickingStudent? = null,
) {
    @Serializable
    data class PickingScreenMeta(
        val mode: String? = null,
        @SerialName("can_pick") val canPick: Boolean? = null,
        @SerialName("session_seed") val sessionSeed: String? = null,
        @SerialName("supported_filters") val supportedFilters: List<String>? = null,
    )

    @Serializable
    data class PickingStudent(
        val days: Int? = null,
        val source: String? = null,
        @SerialName("student_id") val studentId: String? = null,
    )
}

@Serializable
data class PickSection(
    @SerialName("section_id") val sectionId: String,
    val title: String? = null,
    @SerialName("sort_order") val sortOrder: Int? = null,
    @SerialName("filter_counts") val filterCounts: FilterCounts? = null,
    val topics: List<PickTopic>? = null,
) {
    val id: String get() = sectionId
}

@Serializable
data class PickTopic(
    @SerialName("topic_id") val topicId: String,
    val title: String? = null,
    @SerialName("sort_order") val sortOrder: Int? = null,
    val coverage: PickCoverage? = null,
    val progress: PickProgress? = null,
    @SerialName("topic_state") val topicState: PickTopicState? = null,
    @SerialName("filter_counts") val filterCounts: FilterCounts? = null,
) {
    val id: String get() = topicId
}

@Serializable
data class PickCoverage(
    @SerialName("total_proto_count") val totalProtoCount: Int? = null,
    @SerialName("covered_proto_count") val coveredProtoCount: Int? = null,
)

@Serializable
data class PickProgress(
    @SerialName("all_time_pct") val allTimePct: Int? = null,
    @SerialName("last_seen_at") val lastSeenAt: String? = null,
    @SerialName("attempt_count_total") val attemptCountTotal: Int? = null,
    @SerialName("correct_count_total") val correctCountTotal: Int? = null,
    @SerialName("subtopic_last3_avg_pct") val subtopicLast3AvgPct: Double? = null,
)

@Serializable
data class PickTopicState(
    @SerialName("is_stale") val isStale: Boolean? = null,
    @SerialName("is_low_seen") val isLowSeen: Boolean? = null,
    @SerialName("is_not_seen") val isNotSeen: Boolean? = null,
    @SerialName("is_unstable") val isUnstable: Boolean? = null,
)

@Serializable
data class FilterCounts(
    val stale: Int? = null,
    val unstable: Int? = null,
    @SerialName("unseen_low") val unseenLow: Int? = null,
    @SerialName("weak_spots") val weakSpots: Int? = null,
)

// MARK: Статистика прототипов (proto_last3_*, question_stats_for_teacher_*)

/**
 * Бейдж «последние 3 попытки» по базовому прототипу (unic_id);
 * self-версия дополнительно отдаёт total/correct/last_attempt_at.
 */
@Serializable
data class ProtoLast3Stat(
    @SerialName("unic_id") val unicId: String,
    @SerialName("last3_total") val last3Total: Int? = null,
    @SerialName("last3_correct") val last3Correct: Int? = null,
    val total: Int? = null,
    val correct: Int? = null,
    @SerialName("last_attempt_at") val lastAttemptAt: String? = null,
)

@Serializable
data class QuestionStat(
    @SerialName("question_id") val questionId: String,
    val total: Int? = null,
    val correct: Int? = null,
    @SerialName("last3_total") val last3Total: Int? = null,
    @SerialName("last3_correct") val last3Correct: Int? = null,
    @SerialName("last_attempt_at") val lastAttemptAt: String? = null,
)

// MARK: teacher_picking_resolve_batch_v1

@Serializable
data class ResolveBatchResult(
    @SerialName("picked_questions") val pickedQuestions: List<ResolvedQuestion>? = null,
    val shortages: List<ResolveShortage>? = null,
)

@Serializable
data class ResolveShortage(
    @SerialName("scope_id") val scopeId: String? = null,
    @SerialName("is_shortage") val isShortage: Boolean? = null,
    @SerialName("requested_n") val requestedN: Int? = null,
    @SerialName("returned_n") val returnedN: Int? = null,
    val message: String? = null,
)

/**
 * Вопрос из resolve-батча. question_id + topic_id достаточно для
 * создания ДЗ (frozen_questions) и резолва текста через ContentService.
 */
@Serializable
data class ResolvedQuestion(
    @SerialName("question_id") val questionId: String,
    @SerialName("topic_id") val topicId: String? = null,
    @SerialName("proto_id") val protoId: String? = null,
    @SerialName("manifest_path") val manifestPath: String? = null,
    /** Атрибуция бакета из ответа батча (для клиентской ротации по бакетам). */
    @SerialName("scope_kind") val scopeKind: String? = null,
    @SerialName("scope_id") val scopeId: String? = null,
) {
    val id: String get() = questionId
}

// MARK: создание ДЗ (REST insert)

@Serializable
data class HomeworkRow(
    val id: String,
    val title: String? = null,
)

@Serializable
data class HomeworkLinkRow(
    val id: String? = null,
    val token: String? = null,
)
